### PR TITLE
Fix/callable protocols 2

### LINKED ISSUE #595: Callable protocols inheritance broken
## 🐛 Bug report

Callable protocols (implement `__call__`) seems to be parser incorrectly in comparison to "normal" protocols.

### To reproduce

Example that works
```
from dataclasses import dataclass
from typing import Protocol

from jsonargparse import ArgumentParser


class BaseGood(Protocol):
    def predict(self, a: int) -> int: ...


@dataclass
class SpecializedGood(BaseGood):
    number: int = 5

    def predict(self, a: int) -> int:
        return a + 5


def main() -> None:
    parser = ArgumentParser(description="CLI for training models", exit_on_error=False)
    parser.add_argument("--cfg", type=BaseGood)
    args = parser.parse_args(["--cfg=SpecializedGood", "--cfg.number=5"])
    print(args)
    
if __name__ == "__main__":
    main()
```

Example that does not work:
```
class BaseBad(Protocol):
    def __call__(self, a: int) -> int: ...


@dataclass
class SpecializedBad(BaseBad):
    number: int = 5

    def __call__(self, a: int) -> int:
        return a + 5


def main() -> None:
    parser = ArgumentParser(description="CLI for training models", exit_on_error=False)
    parser.add_argument("--cfg", type=BaseBad)
    args = parser.parse_args(["--cfg=SpecializedBad", "--cfg.number=5"])
    print(args)
    
if __name__ == "__main__":
    main()
```

The second example raises the following error:
```
Import path SpecializedBad does not correspond to a subclass of BaseBad
```
Seems like the parser does not match the inherited protocol signature to the base if the protocol methods are private (?).

### Expected behavior

I would expect the second example to work like the first.

### Environment

- jsonargparse version: 4.33.1
- Python version: 3.10
- How jsonargparse was installed: pip
- OS: MacOS

### PR DESCRIPTION
**Pull request by MiguelMonteiro**
## What does this PR do?

Fixes callable protocol inheritance by allowing a list of private methods to be considered when matching signatures for inheritance of subclasses.  The set of `allowed_dunder_methods` currently only contains the `__call__` method.

Fixes #595

## Before submitting

- [x] Did you read the [contributing guideline](https://github.com/omni-us/jsonargparse/blob/main/CONTRIBUTING.rst)?
- [ ] Did you update **the documentation**? (readme and public docstrings)
- [ ] Did you write **unit tests** such that there is 100% coverage on related code? (required for bug fixes and new features)
- [x] Did you verify that new and existing **tests pass locally**?
- [x] Did you make sure that all changes preserve **backward compatibility**?
- [x] Did you update **the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -11,6 +11,13 @@ The semantic versioning only considers the public API as described in
 :ref:`api-ref`. Components not mentioned in :ref:`api-ref` or different import
 paths are considered internals and can change in minor and patch releases.
 
+v4.34.1 (2024-10-??)
+--------------------
+
+Fixed
+^^^^^
+- Fix callable protocol inheritance.
```
Comment (mauvilsa) [inline CHANGELOG.rst:19]: v4.34.0 hasn't been released yet, so this entry should be there instead of a new v4.34.1 section. Also please add a link to this pull request.

File: jsonargparse/_typehints.py [hunk 2]
```diff
@@ -1101,14 +1101,15 @@ def adapt_typehints(
...
+    allowed_dunder_methods = {"__call__"}
     from jsonargparse._parameter_resolvers import get_signature_parameters
     from jsonargparse._postponed_annotations import get_return_type
 
     if not inspect.isclass(value):
         return False
     members = 0
     for name, _ in inspect.getmembers(protocol, predicate=inspect.isfunction):
-        if name.startswith("_"):
+        if name.startswith("_") and name not in allowed_dunder_methods:
```
Comment (mauvilsa) [inline jsonargparse/_typehints.py:1112]: Give me some time and I analyze the dunder methods in general and maybe in this pull request we do more than `__call__`.

**Review by mauvilsa — CHANGES_REQUESTED** [review]
Please add a unit test for a protocol with `__call__`.

**Comment by MiguelMonteiro** [comment]
> Please add a unit test for a protocol with `__call__`.

I added some tests for callable protocols. One question that came up is whether functions should be checked against the protocol. Currently that test is failing, here is the example:
```
class CallableInterface(Protocol):
    def __call__(self, items: List[float]) -> List[float]: ...


class ImplementsCallableInterface1:
    def __init__(self, batch_size: int):
        self.batch_size = batch_size

    def __call__(self, items: List[float]) -> List[float]:
        return items

def implements_callable_interface2(items: List[float]) -> List[float]:
    return items
 ```
 The result is:
 ```
 implements_protocol(ImplementsCallableInterface1, CallableInterface) = True
implements_protocol(implements_callable_interface2, CallableInterface) = False
 ```

File: jsonargparse/_typehints.py [hunk 3]
```diff
@@ -1101,14 +1101,15 @@ def adapt_typehints(
...
     from jsonargparse._parameter_resolvers import get_signature_parameters
     from jsonargparse._postponed_annotations import get_return_type
 
-    if not inspect.isclass(value):
+    if not inspect.isclass(value):  # Should we check if callables implement protocols?
         return False
     members = 0
     for name, _ in inspect.getmembers(protocol, predicate=inspect.isfunction):
-        if name.startswith("_"):
+        if name.startswith("_") and name not in allowed_dunder_methods:
```
Comment (mauvilsa) [inline jsonargparse/_typehints.py:1112]:
```suggestion
        is_dunder = name.startswith("__") and name.endswith("__")
        if (not is_dunder and name.startswith("_")) or (is_dunder and name in protocol_irrelevant_dunder_methods):
```

This is what I came up with.

File: jsonargparse/_typehints.py [hunk 4]
```diff
@@ -1101,14 +1101,15 @@ def adapt_typehints(
 
 
 def implements_protocol(value, protocol) -> bool:
+    allowed_dunder_methods = {"__call__"}
```
Comment (mauvilsa) [inline jsonargparse/_typehints.py:1104]:
```suggestion
protocol_irrelevant_dunder_methods = {
    "__init__", "__new__", "__del__", "__getattr__", "__getattribute__", "__setattr__",
    "__delattr__", "__reduce__", "__reduce_ex__", "__getstate__", "__setstate__"
}

def implements_protocol(value, protocol) -> bool:
```

File: jsonargparse/_typehints.py [hunk 5]
```diff
@@ -1101,14 +1101,15 @@ def adapt_typehints(
...
     from jsonargparse._parameter_resolvers import get_signature_parameters
     from jsonargparse._postponed_annotations import get_return_type
 
-    if not inspect.isclass(value):
+    if not inspect.isclass(value):  # Should we check if callables implement protocols?
```
Comment (mauvilsa) [inline jsonargparse/_typehints.py:1108]: It does make sense that a function could implement a callable protocol. But I would leave this for a different pull request.

**Review by mauvilsa — CHANGES_REQUESTED** [review]

**Comment by MiguelMonteiro** [comment]
Thanks for the feedback, sorry it's been really busy at work, I will try to get to this next week

**Comment by MiguelMonteiro** [comment]
I made the necessary fixes, currently only one subtest is failing but I don't fully understand why.

```
@pytest.mark.parametrize(
    "expected, value",
    [
        (True, ImplementsCallableInterface1),
        (False, ImplementsCallableInterface1(1)),
        (
            False,
            implements_callable_interface2,
        ),  # TODO: switch to True once functions that implement protocol are checked correctly
        (False, NotImplementsCallableInterface1),
        (False, NotImplementsCallableInterface2),
        (False, NotImplementsCallableInterface3),
        (False, not_implements_callable_interface4),
        (False, object),
    ],
)
def test_implements_callable_protocol(expected, value):
    assert implements_protocol(value, CallableInterface) is expected
```

The last subtest ` (False, object)` raises the following error instead of returning `False`

```FAILED jsonargparse_tests/test_subclasses.py::test_implements_callable_protocol[False-object] - ValueError: Invalid or unsupported input: class=<class 'object'>, method_or_property=__call__```

It does not happen in the normal (not callable) protocol tests.

File: jsonargparse/_typehints.py (hunk 5 above)
Comment (MiguelMonteiro) [reply jsonargparse/_typehints.py:1110]: Shall I open an issue so we can keep track of this feature/bug?

Comment (mauvilsa) [reply jsonargparse/_typehints.py:1110]: Sure, open an issue.

File: jsonargparse/_typehints.py [hunk 6]
```diff
@@ -1087,15 +1087,32 @@ def adapt_typehints(
...
+    "__subclasshook__",
+}
+
+
 def implements_protocol(value, protocol) -> bool:
     from jsonargparse._parameter_resolvers import get_signature_parameters
     from jsonargparse._postponed_annotations import get_return_type
 
-    if not inspect.isclass(value):
+    if not inspect.isclass(value):  # Should we check if callables implement protocols?
```
Comment (mauvilsa) [inline jsonargparse/_typehints.py:1110]:
```suggestion
    if not inspect.isclass(value) or value is object:
```
Simple solution to avoid the problem with object.

File: jsonargparse_tests/test_subclasses.py [hunk 7]
```diff
@@ -1500,6 +1500,66 @@ def test_parse_implements_protocol(parser):
...
+
+    def __call__(self, items: List[float]) -> List[float]:
+        return items
+
+
+# TODO: make functions that implement protocol be checked correctly
+def implements_callable_interface2(items: List[float]) -> List[float]:
+    return items
+
+
```
Comment (mauvilsa) [inline jsonargparse_tests/test_subclasses.py:1522]:
```suggestion
```
Add this when it gets implemened.

File: jsonargparse_tests/test_subclasses.py [hunk 8]
```diff
@@ -1500,6 +1500,66 @@ def test_parse_implements_protocol(parser):
...
+@pytest.mark.parametrize(
+    "expected, value",
+    [
+        (True, ImplementsCallableInterface1),
+        (False, ImplementsCallableInterface1(1)),
+        (
+            False,
+            implements_callable_interface2,
+        ),  # TODO: switch to True once functions that implement protocol are checked correctly
```
Comment (mauvilsa) [inline jsonargparse_tests/test_subclasses.py:1551]:
```suggestion
```
Add this when it gets implemened.

**Review by mauvilsa — CHANGES_REQUESTED** [review]
Looks good. But please remove the TODOs as commented.

**Comment by MiguelMonteiro** [comment]
all should be good now

**Review by mauvilsa — APPROVED** [review]
